### DISCUSSION TITLE
Unable to use `overrideProvider` and `useMocker` on CommandTestFactory

### DISCUSSION THREAD

**Discussion by keawade** [jmcdo29/nest-commander#727] [Q&A] [answered]
I'm having trouble with the `CommandTestFactory` tooling. My CLI module has a provider that I need to mock so it doesn't touch my disk during tests but I've yet to be able to get the provider to be defined at all, much less replaced with my mock provider. I'm hoping that I'm just missing something with how overriding providers is supposed to work.

All of these permutations seem to have the same result: The command class property defined by the Provider that is injected into my command has a value of `undefined` when run through the `CommandTestFactory`.

The examples below use this mock class so I should see a log when the mock is constructed.

```typescript
import { Injectable } from '@nestjs/common';
import { vi } from 'vitest';

@Injectable()
export class MyProviderMock {
  public constructor() {
    console.log('MyProviderMock constructor');
  }

  public myMethod = vi.fn();
}
```

```typescript
// Attempting to override the provider with `useClass`
commandInstance = await CommandTestFactory.createTestingCommand({
  imports: [MyModule],
})
  .overrideProvider('MyProvider')
  // The mock's constructor log is not seen with this version
  .useClass(MyProviderMock)
  .compile();
```

```typescript
// Attempting to override the provider with `useFactory`
commandInstance = await CommandTestFactory.createTestingCommand({
  imports: [MyModule],
})
  .overrideProvider('MyProvider')
  .useFactory({
    factory: () => {
      // This log is not seen
      console.log('Hello from the factory');
      // The mock's constructor log is not seen with this version
      return new MyProviderMock();
    }),
  })
  .compile();
```

```typescript
// Attempting to override the provider with `useValue`
commandInstance = await CommandTestFactory.createTestingCommand({
  imports: [MyModule],
})
  .overrideProvider('MyProvider')
  // The constructor's log is seen with this version but the resulting value is still `undefined`
  .useValue(Object.assign({}, new MyProviderMock()))
  .compile();
```

```typescript
// Attempting to just run without overriding at all has the same result
commandInstance = await CommandTestFactory.createTestingCommand({
  imports: [MyModule],
}).compile();
```

```typescript
// Attempting to inject mock with `useMocker`
commandInstance = await CommandTestFactory.createTestingCommand({
  imports: [MyModule],
})
  .useMocker((token) => {
    // No logs here are seen when run
    console.log(`useMocker: token: ${String(token)}`);

    if (token === MyProvider) {
      console.log('useMocker found MyProvider token');
      return { myMethod: vi.fn() };
    }
  })
  .compile();
```

I'm hoping you might have some feedback to point me in a direction to go from here. In the mean time I'm going to start putting together a minimum reproduction repo to more easily sanity check my setup.

**Comment by jmcdo29** [comment, marked as answer]
Do you normally use `@Inject('MyProvider')` and a custom provider with `provide: 'MyProvider'`? If not, you need to use the class reference `MyProvider` instead of the string name of the class

**Reply by keawade** [reply]
'Doh! Yeah I wasn't using `@Inject(MyProvider)` at all. My command classes looked like this:

```typescript
@Command({
  name: 'my-command',
  description: "it's neat, okay?",
})
export class MyCommand extends CommandRunner {
  public constructor(private readonly myProvider: MyProvider) {
    super();
  }

  public async run(): Promise<void> {
    ...
  }
}
```

I'm not quite sure why that even worked at run time outside of the test tooling but now the test tooling works too! Thanks!

**Reply by jmcdo29** [reply]
Actually, what you have and `@Inject(MyProvider)` are essentially the same. They both work because the metadata points to the `MyProvider` _class reference_. What you were doing in your tests was using the token `'MyProvider'`, the string that has the same name as the class, but not the same implementation, which is why the string token and the `overrideProvider` didn't work, because Nest didn't know what `'MyProvider'` meant in terms of the providers in its memeory.

Either way, glad you found your answer :smile:
